Docs for Refreshes and Generation Id (#38780)
airbytehq/airbyte: docs/operator-guides/refreshes.md

---
products: all
---

# Refreshes

:::info

This page describes a coming-soon feature. Refreshes will be available as of Airbyte v1.0.0.

:::

## Refreshes Overview

As of Airbyte 1.0, Refresh Syncs have been added to the Airbyte platform. A Refresh Sync erases the current cursor you have set for any incremental streams, effectively ‘rewinding’ the sync back to the beginning. This is helpful in the following scenarios:

> - The source you syncing from ‘hides’ deleted records, and you wish to truly mirror the source stream, including removing deleted records

Reviewer: ```suggestion
- The source you're syncing from ‘hides’ deleted records, and you wish to truly mirror the source stream, which would include removing deleted records
```

> - The source you syncing from ‘hides’ deleted records, and you wish to truly mirror the source stream, including removing deleted records

Reviewer 2: ```suggestion
- The source you syncing from does not sync deleted records, and you wish to truly mirror the source stream, including removing deleted records
```

Reviewer 2: I think `hides` could be confusing here

Author: 👍

> - The source you're syncing from does not sync record deletions/removals, and you wish to truly mirror the source stream, which would include removing deleted records

Reviewer 3: nit: is it really table mirroring if we have duplicates? (b/c resumable full refresh)

Author: yeah... this is hard to talk about :(.  I'll remove the word 'truly' at least ;D

- The source you are syncing from is unreliable and you wish to re-attempt the sync
- Something has gone wrong with your sync, and you wish to try again

> Refresh syncs do not create data downtime. While the refresh is running, the previous data in your destination remains available to query. Only at the successful conclusion of a refresh sync with the newly refreshed data be swapped or merged into place. This is an improvement from the previous "Reset Syncs" that Airbyte had in earlier iterations.

Reviewer 2: ```suggestion
`Refresh Syncs` do not create data downtime. While the `Refresh Sync` is running, the previous data in your destination remains available to query. Only at the successful conclusion of a `Refresh Sync` will the newly refreshed data be swapped or merged into place. This is an improvement from the previous `Reset Syncs` that Airbyte had in earlier iterations.
```

> When performing a Refresh Sync, any records you’ve already imported will be imported again for the stream(s) being refreshed (assuming they are still present within the source), leading to a slower and more expensive sync than you might be used to. Refresh syncs will block any scheduled syncs from occurring until they are complete. During a Refresh Sync, the streams that are not being refreshed will also sync normally in order to lower data latency . All refresh syncs are resumable, if the underlying source and destination support [checkpointing](/understanding-airbyte/airbyte-protocol/#state--checkpointing).

Reviewer 2: ```suggestion
When performing a `Refresh Sync`, any records you’ve already imported will be imported again for the stream(s) being refreshed (assuming they are still present within the source), leading to a slower and more expensive sync than you might be used to. `Refresh Syncs` will block any scheduled syncs from occurring until they are complete. During a `Refresh Sync`, the streams that are not being refreshed will also sync normally in order to lower data latency . All `Refresh Syncs` are resumable if the underlying source and destination support [checkpointing](/understanding-airbyte/airbyte-protocol/#state--checkpointing).
```

Reviewer 2: Should we use `Refresh Sync` or just Refresh Sync (no code formatting) ?

| Stream Sync Mode               | Refresh stream and remove records | Refresh stream and retain records | Clear     |
| ------------------------------ | --------------------------------- | --------------------------------- | --------- |
| Incremental \| Append + Dedupe | Available                         | Available                         | Available |
| Full Refresh \| Overwrite      | Not available                     | Not available                     | Available |
| Full Refresh \| Append         | Not available                     | Not available                     | Available |
| Incremental \| Append          | Not available                     | Available                         | Available |

For supported destinations, and depending on the sync mode you have selected, there are two types of refreshes - "retaining history" and "removing history". Both kinds of Refresh Syncs can be triggered from the UI and via the Airbyte API at either the stream or connection levels (although both might not be available depending on your sync mode).

![Are you sure you want to refresh data](../.gitbook/assets/refreshes/are-you-sure-refresh.png)

## Refresh and Remove Records

> Refresh and Remove Records Syncs will only display data in the destination’s final tables that was newly synced since the refresh was initiated. This is the simplest type of refresh, and leads to table mirroring, if you are using an overwrite sync mode.

Reviewer 2: > if you are using an overwrite sync mode.

Isn't it table mirroring regardless of sync mode? I think we can remove that line.

Author: No... any append sync mode isn't a mirror as it captures history, or CDC.

Reviewer 2: I think I worded that wrong. What i meant was - running `Refresh and Remove` on a stream will lead to table mirroring, regardless of sync mode. Right?

Author: Probably... but maybe things get weird with CDC or resumable syncs... either way I removed it.

:::warning

> Not all sources keep their history forever. If you perform a Refresh and Remove Records Sync, and your source does not retain all of its records, this may lead to data loss.

Reviewer 2: ```suggestion
Not all sources keep their history forever. If you perform a Refresh and Remove Records Sync, and your source does not retain all of its records, this may lead to data loss. If you need to keep all history, then we recommend using `Refresh and Retain Records`. If that is not an option, then you should back up the data in your destination before running the Refresh and Remove Records.
```

:::

## Refresh and Retain Records

> Refresh and Retain Records Syncs keep the records you had previously, and merge in the new data from this sync onward with the data previously stored in your destination. Using a Refresh and Retain Records Sync with a Deduplicated Sync mode is the best way to recover from an unreliable source.

Reviewer 2: Have we defined what an `unreliable source` is anywhere? If not, we should, and link to it here. I'm not sure it will be a widely understood term.

Author: I don't think so. I added `(e.g. lossy or buggy)`

:::warning

Refresh and Retain Records Syncs will lead to increased data storage in your destination (as the old and new copies of your data are both stored). This may also lead to increased compute costs and slower syncs, depending on your destination.

:::

## Resumability

> Airbyte strives to create resilient syncs that can handle many types of issues (for example, networking issues, flaky source APIs or under provisioned destination servers). Whenever possible, Airbyte syncs are resumable, including standard Full Refresh + Overwrite and Full Refresh + Append syncs, along with the Refreshes we are describing on this page. “Resumable” in this context means that if something goes wrong in the first attempt of your sync, we will immediately try again with a subsequent attempt - in fact, Airbyte will keep retrying as long as the source is producing new records and the destination can commit them. In order to achieve this resumability, Airbyte may deliver a record more than once - Airbyte is a deliver at-least-once system.

Reviewer 4: This will be available soon after refreshes. Do we want to make it explicit in the docs for now?

Author: Is this different from what the admonation at the top says (This page describes a coming-soon feature. Refreshes will be available as of Airbyte v1.0.0.)?

Reviewer 4: They both should be available by 1.0.0. But technically, since we will be releasing Refresh before RFR, it could be confusing for a user.

This example uses a database source, but the same logic holds for API sources as well. A series of requests to a users API endpoint might be:

- `curl -x GET api.com/v1/users?page=1`
- `curl -x GET api.com/v1/users?page=2`
- `...`

## Data Generations

With the advent of Refresh and Retain History Syncs, Airbyte has provided a way for you to determine if any given record is from before or after the refresh took place. This is helpful when disambiguating historic data which may have changed due to an unreliable source. We call this information the “Generation” of the data, indicated by the `_airbyte_generation_id` column. Every time a Refresh Sync (of either type) occurs, the generation increases. `_airbyte_generation_id` is a monotonically increasing counter which maps to this notion of “generation” of data.

> ### Example: Understanding and Recovering from a Flaky Source

Reviewer 2: nit: I think generation starts at 0 instead of 1
